geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala

```diff
@@ -0,0 +1,224 @@
+
+  def encodeDeleteMessage(topic: String, msg: Delete): MSG = {
+    val key = encodeKey(msg)
+    val value = msg.id.getBytes(StandardCharsets.UTF_8)
+    new MSG(topic, key, value)
+  }
+}
+
+/** For encoding any [[KafkaGeoMessage]].
+  *
+  * @param schema the [[SimpleFeatureType]]; required to serialize [[CreateOrUpdate]] messages
+  */
+class KafkaGeoMessageEncoder(val schema: SimpleFeatureType) {
+
+  import KafkaGeoMessageEncoder.{MSG, encodeKey}
+
+  val sfEncoder: SimpleFeatureEncoder = new AvroFeatureEncoder(schema, EncodingOptions.withUserData)
+
+  def encode(topic: String, msg: KafkaGeoMessage): MSG = msg match {
+    case c: CreateOrUpdate =>
+      encodeCreateOrUpdateMessage(topic, c)
+    case d: Delete =>
+      encodeDeleteMessage(topic, d)
+    case x: Clear =>
+      encodeClearMessage(topic, x)
+  }
+
+  val encodeClearMessage: (String, Clear) => MSG = KafkaGeoMessageEncoder.encodeClearMessage
+
+  val encodeDeleteMessage: (String, Delete) => MSG = KafkaGeoMessageEncoder.encodeDeleteMessage
+
+  def encodeCreateOrUpdateMessage(topic: String, msg: CreateOrUpdate): MSG = {
+    val key = encodeKey(msg)
+    val value = sfEncoder.encode(msg.feature)
+    new MSG(topic, key, value)
+  }
+}
+
+class KafkaGeoMessageDecoder(val schema: SimpleFeatureType) extends Logging {
+
+  type MSG = MessageAndMetadata[Array[Byte], Array[Byte]]
+
+  private lazy val V0_DELETE_KEY = "delete".getBytes(StandardCharsets.UTF_8)
+  private lazy val V0_CLEAR_KEY = "clear".getBytes(StandardCharsets.UTF_8)
+  private lazy val V0_TIMESTAMP = new Instant(0L)
+
+  val sfDecoder: SimpleFeatureDecoder = new AvroFeatureDecoder(schema, EncodingOptions.withUserData)
+
+  def decode(msg: MSG): KafkaGeoMessage = {
+
+    val key = msg.key()
+
+    if (key != null && key(0) == 1) {
+      decodeVersion1(msg)
+    } else {
+      // try version 0
+      decodeVersion0(msg)
+    }
+  }
+
+  def decodeKey(key: Array[Byte]): (Instant, Byte) = {
+    if (key.length != 10) {
+      throw new IllegalArgumentException(s"Invalid message key.  Expecting 10 bytes but found ${key.length}: "
+        + s"${new String(key, StandardCharsets.UTF_8)}")
+    }
+
+    val buffer = ByteBuffer.wrap(key)
+
+    buffer.get() // version
+    val msgType = buffer.get()
+    val ts = new Instant(buffer.getLong)
+
+    (ts, msgType)
+  }
+
+  def decodeVersion1(msg: MSG): KafkaGeoMessage = {
+    val (ts, msgType) = decodeKey(msg.key())
+
+    msgType match {
+      case KafkaGeoMessageEncoder.createOrUpdateType =>
+        val sf = sfDecoder.decode(msg.message())
+        CreateOrUpdate(ts, sf)
+      case KafkaGeoMessageEncoder.deleteType =>
+        val id = new String(msg.message(), StandardCharsets.UTF_8)
+        Delete(ts, id)
+      case KafkaGeoMessageEncoder.clearType =>
+        Clear(ts)
+      case _ =>
+        throw new IllegalArgumentException("Unknown message type: " + msgType.toChar)
+    }
+  }
+
+  def decodeVersion0(msg: MSG): KafkaGeoMessage = {
```

Comment:
I don't think there's reason to keep backwards compatibility at this point . . . but keep the version in the meta data for future backwards compatibility that *is* necessary.